cmd/soci/commands/internal/index.go

```diff
@@ -0,0 +1,40 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package internal
+
+import "github.com/urfave/cli"
+
+const (
+	ExistingIndexFlagName = "existing-index"
+	Warn                  = "warn"
+	Abort                 = "abort"
```

Comment:
SOCI does support pushing multiple indices in the sense that you can push multiple platforms with one command, but I think your overall statement still stands.

I think the real use-case for `abort` is that it exits non-zero if an index already exists.

Reply:
> I think the real use-case for abort is that it exits non-zero if an index already exists.

That can be part of `warn` if needed?

I think `abort` in general doesn't provide a good UX. It means a user need to go back-and-forth even to only complete the command for those valid inputs. 

It's more common for CLIs to just log/accumuate the errors/warns and proceed to the next input, at least in docker/nerdctl (e.g., `docker rmi -f $(docker images -q)` and `docker rm -f $(docker ps -q)`). Then upon complete, the user know which valid inputs are processed and, from the warn/error, which invalid inputs are left and figure out what they want to do with them (e.g., `skip` or `allow`)

Reply:
> The case for abort that I see is when you're scripting pushing a single index. With multiple indices, I agree that skip would make more sense.

Then I'd suggest remove `abort` now, given that we expect this command to be used with multiple indices. If we really see a usecase we can always add it back.

As an example, `docker rmi` won't `abort` even without `-f`. (`-f` is just used to decide log error or return error with non-0 code).

https://github.com/docker/cli/blob/881c3535760afdf9eecb97f34da2983d0d91af24/cli/command/image/remove.go#L63-L79

Reply:
Yeah I suppose there isn't a real **need** for abort, although it could be a nice to have when debugging/playing around with the CLI. If we plan on adding an extra args eg: `replace` which imo has more of a usecase then I'm fine with removing `abort`.

Reply:
WDYT: @Kern-- @djdongjin

Reply:
I still prefer removing `abort` for now as we don't see a practical usecase for users. For debugging, I think `skip` is sufficient as it also log the skipped manifest. Even if we want to have a non-zero exit code option, its behavior should be the same as `skip` (e.g. log/skip and proceed) except that the exit-code is non-0 and the message is error instead of log. This follows docker/nerdctl cli's `-f` flag (without `-f`, skip and return all errors; with `-f`, skip and print logs)

`replace` is nice-to-have and if/when we need it we can add in follow-up PRs.

Reply:
I'm not meant to add a `-f` for `skip` :). Just take, e.g., `docker/nerdctl rmi -f` as an example for non-zero exit scenario: if there is any error, `docker rmi img1 img2` will return non-zero whereas `docker rmi -f rmi1 img2` return 0, and regardless w or w/o `-f`, it always `skip` and proceed.

Still I don't feel `abort` provides a good UX except maybe the scenario you mentioned (In the single platform case for non-zero exit). The thing about `abort` is that it has random behavior (e.g., `cmd --abort good-1, good-2, bad-1` triggers the cmd on good-1 and good-2, but `cmd --abort good-1, bad-1, good-2` only triggers on good-1). And I haven't seen a `abort in the middle` case in common clis. Some counter-example includes `docker/nerdctl rmi` and `rm`, both just skip and proceed and return non-zero at the end:

```shell
➜ ~ touch a b
➜ ~ mkdir c-dir
➜ ~ rm c-dir a b
rm: cannot remove 'c-dir': Is a directory
# 1 exit code
➜ ~ echo $?
1
# still both a and b are deleted
➜ ~ ls
c-dir
```

Probably we should have `warn,skip,allow`:

1. `allow`: push, return 0.
2. `skip`: not push, return 0, print log which are not pushed.
3. `warn`: not push, return 0 or non-0 depending if any input has existing index, print warn/error which are not pushed.

Reply:
Yes, I think the right `abort` semantics for multi-platform indices is to check if any platform already has an index in the registry __before pushing anything__ and exit with non-zero without pushing anything if that check fails. That's not what we had implemented and I can agree that it's probably not a huge usecase to worry about.